posthog/temporal/batch_exports/monitoring.py

```diff
@@ -0,0 +1,220 @@
+        if not await client.is_alive():
+            raise ConnectionError("Cannot establish connection to ClickHouse")
+
+        response = await client.read_query(query, query_params)
+        results = []
+        for line in response.decode("utf-8").splitlines():
+            interval_start, interval_end, count = line.strip().split("\t")
+            results.append(
+                EventCountsOutput(interval_start=interval_start, interval_end=interval_end, count=int(count))
+            )
+
+        return GetEventCountsOutputs(results=results)
+
+
+@dataclass
+class UpdateBatchExportRunsInputs:
+    batch_export_id: UUID
+    results: list[EventCountsOutput]
+
+
+@activity.defn
+async def update_batch_export_runs(inputs: UpdateBatchExportRunsInputs) -> int:
+    """Update BatchExportRuns with the expected number of events."""
+
+    total_rows_updated = 0
+    for result in inputs.results:
+        total_rows_updated += await aupdate_expected_records_count(
+            batch_export_id=inputs.batch_export_id,
+            interval_start=dt.datetime.strptime(result.interval_start, "%Y-%m-%d %H:%M:%S").replace(tzinfo=dt.UTC),
+            interval_end=dt.datetime.strptime(result.interval_end, "%Y-%m-%d %H:%M:%S").replace(tzinfo=dt.UTC),
+            count=result.count,
+        )
+    activity.logger.info(f"Updated {total_rows_updated} BatchExportRuns")
+    return total_rows_updated
+
+
+@workflow.defn(name="batch-export-monitoring")
+class BatchExportMonitoringWorkflow(PostHogWorkflow):
+    """Workflow to monitor batch exports.
+
+    We have had some issues with batch exports in the past, where some events
+    have been missing. The purpose of this workflow is to monitor the status of
+    batch exports for a given customer by reconciling the number of exported
+    events with the number of events in ClickHouse for a given interval.
+    """
+
+    @staticmethod
+    def parse_inputs(inputs: list[str]) -> BatchExportMonitoringInputs:
+        """Parse inputs from the management command CLI."""
+        loaded = json.loads(inputs[0])
+        return BatchExportMonitoringInputs(**loaded)
+
+    @workflow.run
+    async def run(self, inputs: BatchExportMonitoringInputs):
+        """Workflow implementation to monitor batch exports for a given team."""
+        # TODO - check if this is the right way to do logging since there seems to be a few different ways
+        workflow.logger.info("Starting batch exports monitoring workflow for team %s", inputs.team_id)
+
+        batch_export_details = await workflow.execute_activity(
+            get_batch_export,
+            inputs.team_id,
+            start_to_close_timeout=dt.timedelta(minutes=1),
+            retry_policy=RetryPolicy(
+                maximum_attempts=3,
+                initial_interval=dt.timedelta(seconds=20),
+                non_retryable_error_types=["NoActiveBatchExportsFoundError"],
+            ),
+            heartbeat_timeout=dt.timedelta(minutes=1),
```

Comment:
This activity does not heartbeat, so having a heartbeat timeout is pointless.

Reply:
Same thing with other activities: I'd add heartbeating (best) or remove the timeout as it is confusing (sufficient).

Reply:
ah yes, oops, I just copy/pasted this from other jobs without considering that. Will look into adding a hearbeat